feat: create `await new ParseServer(config).startApp()` to resolve cloud code and schema before express app is exposed

dblythy opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Currently, there is no way to use cloud code if you are using ES Modules due to the fact that `import` needs to be async.

Related issue: #7560
Closes: #7527, #7560

### Approach
<!-- Add a description of the approach in this PR. -->

Adds `await new ParseServer().startApp()` that create a Parse Server asynchronously so that a developer using ES imports and schemas can be sure that their cloud triggers and schemas are registered before creating the express app.

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [x] Add tests
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

mtrezza committed:
ci: fix node engine check (#7891)

dblythy committed:
fix: return correct response when revert is used in beforeSave (#7839)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.1 [skip ci]

## [5.2.1-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.0...5.2.1-alpha.1) (2022-03-26)

### Bug Fixes

* return correct response when revert is used in beforeSave ([#7839](https://github.com/parse-community/parse-server/issues/7839)) ([f63fb2b](https://github.com/parse-community/parse-server/commit/f63fb2b338c908f0e7a648d338c26b9daa50c8f2))

dblythy committed:
perf: reduce database operations when using the constant parameter in Cloud Function validation (#7892)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.2 [skip ci]

## [5.2.1-alpha.2](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.1...5.2.1-alpha.2) (2022-03-26)

### Performance Improvements

* reduce database operations when using the constant parameter in Cloud Function validation ([#7892](https://github.com/parse-community/parse-server/issues/7892)) ([48bd512](https://github.com/parse-community/parse-server/commit/48bd512eeb47666967dff8c5e723ddc5b7801daa))

github-actions[bot] committed:
feat: add MongoDB 5.1 compatibility (#7682)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.1 [skip ci]

# [5.3.0-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.2...5.3.0-alpha.1) (2022-03-27)

### Features

* add MongoDB 5.1 compatibility ([#7682](https://github.com/parse-community/parse-server/issues/7682)) ([90155cf](https://github.com/parse-community/parse-server/commit/90155cf1680e5e0499b0000e071c6cb0ce3aef96))

mtrezza committed:
fix: security upgrade parse push adapter from 4.1.0 to 4.1.2 (#7893)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.2 [skip ci]

# [5.3.0-alpha.2](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.1...5.3.0-alpha.2) (2022-03-27)

### Bug Fixes

* security upgrade parse push adapter from 4.1.0 to 4.1.2 ([#7893](https://github.com/parse-community/parse-server/issues/7893)) ([ef56e98](https://github.com/parse-community/parse-server/commit/ef56e98ef65041b4d3b7b82cce3473269c27f6fd))

mtrezza committed:
feat: add MongoDB 5.2 support (#7894)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.3 [skip ci]

# [5.3.0-alpha.3](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.2...5.3.0-alpha.3) (2022-03-27)

### Features

* add MongoDB 5.2 support ([#7894](https://github.com/parse-community/parse-server/issues/7894)) ([6b4b358](https://github.com/parse-community/parse-server/commit/6b4b358f0842ae920e45652f5e8b2afebc6caf3a))

dblythy committed:
fix: allow ES import for cloud string if package type is module

dblythy committed:
revert helper

mtrezza commented:
Restarted failed tests

snyk-bot committed:
refactor: upgrade mongodb from 4.3.1 to 4.4.1 (#7906)

github-actions[bot] committed:
ci: bump environment (#7907)

Moumouls committed:
refactor: add missing schema definitions (#7917)

mtrezza commented on src/ParseServer.js:
```diff
@@ -323,6 +329,25 @@ class ParseServer {
+  /**
+   * @static
+   * Create an async parse server
+   * @param {ParseServerOptions} options the parse server initialization options */
```
We should add comment linting finally, in the meantime could you move the `*/` into the next line?

dblythy committed:
fix: custom database options are not passed to MongoDB GridFS (#7911)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.4 [skip ci]

# [5.3.0-alpha.4](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.3...5.3.0-alpha.4) (2022-04-04)

### Bug Fixes

* custom database options are not passed to MongoDB GridFS ([#7911](https://github.com/parse-community/parse-server/issues/7911)) ([a72b384](https://github.com/parse-community/parse-server/commit/a72b384f76137a3d83ffb69f65cb25aff1bbab4f))

dblythy commented:
I think this might close #7527

dblythy committed:
Update ParseServer.js

snyk-bot committed:
refactor: upgrade @apollo/client from 3.5.8 to 3.5.9 (#7921)

dependabot[bot] committed:
fix: security upgrade moment from 2.29.1 to 2.29.2 (#7931)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.5 [skip ci]

# [5.3.0-alpha.5](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.4...5.3.0-alpha.5) (2022-04-09)

### Bug Fixes

* security upgrade moment from 2.29.1 to 2.29.2 ([#7931](https://github.com/parse-community/parse-server/issues/7931)) ([6b68593](https://github.com/parse-community/parse-server/commit/6b68593eaec17e8b183899d2b92699c9ede7625b))

mtrezza committed:
Merge branch 'alpha' into await-cloud-code

marvinroger committed:
fix: peer dependency mismatch for GraphQL dependencies (#7934)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.6 [skip ci]

# [5.3.0-alpha.6](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.5...5.3.0-alpha.6) (2022-04-11)

### Bug Fixes

* peer dependency mismatch for GraphQL dependencies ([#7934](https://github.com/parse-community/parse-server/issues/7934)) ([b7a1d76](https://github.com/parse-community/parse-server/commit/b7a1d7617b4bcac677cecedfeb6ac4a27447083b))

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.2 to 8.2.4 (#7932)

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.4 to 8.2.6 (#7935)

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.6 to 8.2.7 (#7941)

Moumouls commented on src/ParseServer.js:
```diff
@@ -323,6 +329,26 @@ class ParseServer {
+   * Create an async parse server
+   * @param {ParseServerOptions} options the parse server initialization options 
+   */
+  static createAsync(options) {
+    return new Promise((resolve, reject) => {
+      const parseServer = new ParseServer({
+        ...options,
+        serverStartComplete(error) {
+          if (error) {
+            reject(error);
```
I know that the original use case of `serverStartComplete` will have less sense here.
`serverStartComplete` is still a valid option for an `startAsync`  parse-server.

You need to try to call the original "serverStartComplete()" before on reject or resolve, if `options. serverStartComplete` exists.

Then developers will not have weird behavior, (currently you overwrite the serverStartComplete) if they use startAsync.

Moumouls commented on src/ParseServer.js:
```diff
@@ -323,6 +329,26 @@ class ParseServer {
+   * Create an async parse server
+   * @param {ParseServerOptions} options the parse server initialization options 
+   */
+  static createAsync(options) {
```
not sure of "create". We already have `ParseServer.start()`

I can suggest to use `ParserServer.startAsync()`. I'll also suggest to use JS doc to explain the clear difference between `start` and `startAsync`.

I think we should start to deprecate `start()`. It could save a lot of time to some new developers if they try to perform some code like:

```js
const server = ParseServer.start()
const query = new Parse.Query('XXXX')
// Query could fail since parse-server could be not totally initialized,
// it will be random and depend of network/cpu.
// it could be hard to debug for new developers.
```

start: will return a parse-server app not fully initialized
startAsync: will resolve once parse-server is fully initialized

dblythy replied:
start is used to create a Parse Server with an express app, I suggest using `create` which can be used to attach your own express app

dblythy replied:
I have created the function `startApp` which starts Parse Server and returns the express app. Usage is:

`new ParseServer(config).startApp()`

Moumouls commented on spec/CloudCode.spec.js:
```diff
@@ -39,6 +40,34 @@ describe('Cloud Code', () => {
+  it('should wait for cloud code to load', async () => {
+    const config = Config.get('test');
+    const initiated = new Date();
+    const parseServer = await ParseServer.createAsync({
+      ...config,
+      async cloud() {
+        await new Promise(resolve => setTimeout(resolve, 1000));
```
Add a new test for "startAsync" to cover that we still support `startServerComplete` and we do not overwrite the function internally.

dplewis committed:
refactor: upgrade body-parser from 1.19.1 to 1.19.2 (#7944)

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.7 to 8.2.8 (#7945)

snyk-bot committed:
fix: security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 (#7948)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.7 [skip ci]

# [5.3.0-alpha.7](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.6...5.3.0-alpha.7) (2022-04-25)

### Bug Fixes

* security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 ([#7948](https://github.com/parse-community/parse-server/issues/7948)) ([20fc4e2](https://github.com/parse-community/parse-server/commit/20fc4e23b53c91aac657f894bd70d049b7525c37))

mtrezza committed:
refactor: upgrade winston from 3.5.1 to 3.7.2 (#7954)

dblythy committed:
rename method to create

dblythy committed:
resolve .app

dblythy committed:
Merge branch 'alpha' into await-cloud-code

dblythy commented:
This PR also resolves the issue of the server being exposed before schemas are loaded.

Before, migrations happen once express routes are exposed:

![Screen Shot 2022-04-28 at 8 32 08 pm](https://user-images.githubusercontent.com/4974769/165735292-fdc34c02-92d5-461e-9aa4-3908f26ecf25.png)

After, migrations happen before express routes are exposed:

![Screen Shot 2022-04-28 at 8 40 10 pm](https://user-images.githubusercontent.com/4974769/165735386-f8cf08ab-7262-4010-b78d-ebc29b1e73a8.png)

dblythy committed:
Update index.spec.js

dblythy committed:
create startApp function

dblythy committed:
Update index.spec.js

dblythy committed:
increase coverage

dblythy committed:
Update DefinedSchemas.js

dblythy committed:
close servers

dblythy committed:
Update index.js

dblythy committed:
fix tests

mtrezza commented:
Is this a fix or a feature or a feature + fix included? If its feat + fix, then the feat is a higher order so this would be merged as feat.

If I remember correctly, we've reverted a PR some months ago that was also related to Cloud Code module import. Has the reason for reverting been addressed in this PR?

dblythy commented:
It’s a bit of everything really. And yep, I believe the race condition is fixed by this PR. @Moumouls might have more to say though

Moumouls commented:
Hi there @mtrezza  @dblythy,

> This PR also resolves the issue of the server being exposed before schemas are loaded.

Perfect, it was also a point that I planned to check to support correctly serverless env.

Here I only have one question; if you expose the app port AFTER the migration/init/serverStartComplete, what happens if a developer uses the Parse JS SDK into these functions (before/after migration, startServerComplete). From my knowledge, if I remember correctly the Parse JS SDK onboarded on parse-server use the "serverUrl" to interact with parse-server. So a developer could be blocked (also if he use the rest/graphql api in these functions).

To avoid this issue on my side for serverless env, I found that the we should introduce a new "publicPort" option like the "publicServerUrl" to fix the issue, and only expose parse-server to external traffic once it's ready.

We should not use a "directAccess" strategy, since developers could use GraphQL API or the REST API in these initialization functions.

If you confirm with a test that Parse JS SDK (or rest call, or Graphql call) query into before/after migration /startServerComplete fails, your changes will be a breaking change. And we will need to add a new "publicPort" to expose the completely initialized parse-server to the external traffic.

Tell what do you think about this @dblythy  🙂

Moumouls commented on src/SchemaMigrations/DefinedSchemas.js:
```diff
@@ -171,9 +173,8 @@ export class DefinedSchemas {
   async createDeleteSession() {
-    const session = new Parse.Session();
-    await session.save(null, { useMasterKey: true });
-    await session.destroy({ useMasterKey: true });
+    const { response } = await rest.create(this.config, Auth.master(this.config), '_Session', {});
+    await rest.del(this.config, Auth.master(this.config), '_Session', response.objectId);
   }
```
Owww I see here @dblythy  that you encountered the error since the SDK uses the port.

Moumouls replied:
the old code should not be changed, if the old code works, any Parse SDK/GraphQL/REST call during the init will work.

Moumouls commented on src/ParseServer.js:
```diff
@@ -83,45 +83,72 @@ class ParseServer {
+        resolve(this.app);
+        return;
+      }
+      this.startCallbackSuccess = resolve;
+      this.startCallbackError = reject;
```
Clever 👍

Moumouls commented on src/ParseServer.js:
```diff
@@ -83,45 +83,72 @@ class ParseServer {
   }
 
+  /**
+   * Starts the Parse Server to be served as an express app
```
Add a comment that this fn only resolves once parse-server is fully started (after serverStartComplete)

You can also says that once this fn resolves it means that parse-server is ready to accept external traffic

Moumouls commented on src/ParseServer.js:
```diff
@@ -83,45 +83,72 @@ class ParseServer {
+          }
+        }
         if (serverStartComplete) {
           serverStartComplete();
```
use now "await Promise.resolve(serverStartComplete())" to handle correctly a async serverStartComplete

dblythy commented:
The old code works fine because the core tests don’t care for race conditions / deploying schema before exposure and checking cloud code, and use the current function ParseServer.start. This results in Parse mounting on its express app as soon as it’s created so any endpoint is available, before schemas, cloud code etc, which is problematic in envs which re-scale often. If you have a beforeSave or beforeFind hook that ensures data security and Parse exposes that route before the trigger is registered, then it can lead to big leaks if new instances are regularly deployed. 

This new feature isn’t a breaking change as existing functionality works fine, if you want Parse Server to wait to init before your express app you can `await` the new method, otherwise you can leave as is. No migration is required, only recommended for scaling deployments. 

Any JS SDK methods will only work when .listen is called, but if someone is going to the effort to use `await new ParseServer().startApp()`, I would argue they would expect it to await and only be available when the mount it to their express app. 

My thoughts is that there isn’t any downside to the new function, there will be some improvements to be made for sure but I don’t see why it can’t operate as is alongside the current implementation. Anyone who would implement it would be aware that the server won’t be available until they mount it with express.

dblythy committed:
Update ParseServer.js

dblythy committed:
Update ParseServer.js

Moumouls commented:
> The old code works fine because the core tests don’t care for race conditions

Yes but the new start-up strategy will not expose parse-server port until startServerComplete for example. In my case for example I've some Rest/GraphQL calls during serverStartComplete to seed/perform DB checks before exposing the app.

Also, users that want to use the before/afterMigration function and try to request some data will be blocked.

Parse Server historically has 2 serverUrl properties, one for public traffic, and another one for internal traffic. serverUrl and publicServerUrl. I think we can introduce the same strategy with "publicPort"

You can easily call listen 2 times on express app :)

before/afterMigration, serverStartComplete function without the possibility to call Parse REST/GraphQL/SDK API seems useless to me.


Also, you can test the breaking change by trying to call a Parse.Query into the serverStartComplete. it should fail currently.

mtrezza commented:
> In my case for example I've some Rest/GraphQL calls during serverStartComplete to seed/perform DB checks before exposing the app.

Could you give an example? Does this mean you want to be able to call the Parse Server API before the server has finished its start-up procedure?

dblythy committed:
add holdPublicRoutes

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Merge branch 'alpha' into await-cloud-code

dblythy commented:
> Does this mean you want to be able to call the Parse Server API before the server has finished its start-up procedure?

I'm not quite sure how this is at all possible in the current version, considering in order to call the Parse API the public routes have to be available, and `app.listen` is called. If any "startup" procedures currently require the JS API, then all API routes are publicly available (as Parse Server has been mounted).

To circumvent this, I have added a Parse Server option `holdPublicRoutes`. If this is set to `true`, the Parse API will require a valid `masterKey` until `mountPublicRoutes()` is called. This allows for:

```js
const config = {
   ...
  holdPublicRoutes: true,
}
const parseServer = await new ParseServer(config).startApp();
const app = express();
app.use('/parse', parseServer);
const server = app.listen(12701);
await longStartupMethod();
parseServer.mountPublicRoutes(); // any client API calls before this will return unauthorized: master key is required
```

dblythy committed:
definitions

Moumouls commented:
@dblythy @mtrezza I apologize. I think I found the misunderstanding here 🙂 

From the beginning, on my side, I was talking about the `static start()` ParseServer method. The static start method calls the app.listen in the same NodeJS event loop as the ParseServer constructor. This behavior exposes the app before the server is initialized and allows a developer to perform queries on an initializing server.

The static start method should be maybe refactored/deprecated in favor of a static async start, to only listen once the server is initialized. The developers can await the start or use serverStartComplete.
In the static async start use case, the listen SHOULD occur, after before `if (serverStartComplete) {`. Then developers will be able to perform queries in "serverStartComplete".

I use the static start method on my projects combined with serverStartComplete: ex: https://github.com/Moumouls/parse-next-mono-starter/blob/65d2a9e6d485185c1164619e17d64d4d45034d88/packages/back/src/server.ts#L75

I was also wrong to want parse-server to be able to handle readiness by itself. I think we should not implement options like `mountPublicRoutes();` or `publicPort`. After a quick check of some process app managers/orchestrators, like PM2, Kubernetes, Knative, GCP Cloud Run, and other hosting services, readiness probes are handled very differently. It's out of the scope of parse-server to be able to emit/handle correctly these readiness probes, public traffic routing availability.

Finally, sorry for the wasted time here. I can suggest removing the `mountPublicRoutes` options, and to add the `static` version of `startAsync` with the call to `listen()` BEFORE the startServerComplete in case of a startup initiated from the `static` call (this is the current behavior of parse-server). Then developers will be able to implement their readiness probes, code business logic, after the `await startAsync()` or in the `serverStartComplete`

From the static start and startAsync method the events should occur in this order from my point of view:
- performInitialization
- hooksController.load()
- await new DefinedSchemas(schema, this.config).execute();
- (in case of static start: call app.listen here)
- await Promise.resolve(serverStartComplete())
- startCallbackSuccess()

Then we should be good to go 🚀

mtrezza commented:
> It's out of the scope of parse-server to be able to emit/handle correctly these readiness probes, public traffic routing availability.

There are two sides to readiness:

1) the sender (eg. load balancer) should not send until the recipient (Parse Server) is ready.

2) the recipient (Parse Server) should not process any requests until it has finished its start-up procedure.

For (1) we already have the the `/health` endpoint, and it is definitely within Parse Server's (or the external express route's) scope that the endpoint handles requests that are sent before the server is ready, it should for example not crash or cause undefined behavior.

For (2) I don't think Parse Server should handle *any* type of request until is has started up, unless we want to implement a distinct "start-up" API, which would likely be a more complex discussion and I don't think is currently part of this discussion.

More generally, polling Parse Server for readiness is actually a somewhat outdated technique. The start-of-the-art mechanism would be to launch Parse Server with a configuration of hooks that are called once it's ready / failed / etc, and then the load balancer starts sending requests, restarts or replaces the instance. Also in this case, readiness management is very much within the scope of Parse Server.

mtrezza committed:
Merge branch 'alpha' into await-cloud-code

Moumouls commented:
> Also in this case, readiness management is very much within the scope of Parse Server.

I think Parse Server should provide sufficient tools to handle a graceful start. But for example, PM2 can send traffic once it detects a `server.listen()`, K8 has a flexible readiness probe system, and in the case of parse-server, sending a query to `/health` will work perfectly. In Knative, Google cloud run, the traffic is also sent once the target port configured on GCP is listening.

I think the core problem is how to allow developers to use SDK/REST/GQL API in before/after migration and serverStartComplete function given that these functions should be called BEFORE `listen` to ensure a graceful start.

Or developers need to start a custom express app, with a dedicated port to handle on their side the graceful start with their related business logic ?

dblythy committed:
Merge branch 'alpha' into await-cloud-code

dblythy committed:
Merge branch 'alpha' into await-cloud-code

dblythy commented:
I'm not sure what's pending here or what the general feedback was. This approach covers the following objectives:

- ES imports for cloud code as server won't be mounted to `/parse` before `await .startApp()` is called
- Allows developers optional parameter to run custom initialization logic (such as building schema, custom database ops, or anything else) before mounting routes
- Converts schema initialization to use direct methods (e.g `data.create`) rather than unnecessarily routing through express via the Parse JS SDK

dblythy commented:
Closing as this will be introduced as a breaking change in V6.
